### ISSUE TITLE
Instance context

### ISSUE THREAD

**Issue by samsch** [tgriesser/knex#3047]
# Feature discussion / request

1. Explain what is your use case

Debugging in an http server. I want to be able to add query info (e.g., SQL and response data) to per-request debugging logs. At the moment the only way to do this would be manually adding a handler of some kind to each query (or at least calling `.queryContext()`). Ideally the feature would not require changing the usage of the knex object at all, and where this feature is needed, you would just pass a special instance of the knex object.

2. Explain what kind of feature would support this

A couple possibilities:

A. A device for creating a clone of the knex object, but the event listeners attached to it would *only* receive events for queries created using this cloned knex object (or also possibly knex objects cloned from it? (recursive)).

B. A way to clone the knex object which would allow you to set an ID or reference which could then be used in the global event listeners to identify what knex clone the query originated from.

3. Give some API proposal, how the feature should work

```js
const knex = Knex({ ...config });
// A.
const knexClone = knex.createInstance();

knexClone.on('query', (data) => { console.log(data) });

knexClone('user').select('name').then(...);
// Calls event listener above.

knex('user').select('name').then(...);
// Does *not* call event listener above.

// B.
const instance = {};
const knexClone = knex.createInstance(instance);

knex.on('query', (data) => { console.log(data.instance === instance ? 'child query' : 'other query') });

knexClone('user').select('name').then(...);
// Logs 'child query'

knex('user').select('name').then(...);
// Logs 'other query'
```

**Comment by kibertoad** [comment]
@samsch Have you tried withUserParams() function on knex instance?

**Comment by samsch** [comment]
I did attempt to use it, yes, but it didn't appear to be included in the
event (`query`, `query-response`) data.

**Comment by kibertoad** [comment]
Sounds like a bug. Could you create a failing test for sqlite with your use-case? Then I could look into it.

**Comment by kibertoad** [comment]
You can check out existing tests for withUserParams as reference.

**Comment by elhigu** [comment]
@samsch please add bugreport with reproduction code.  Also with node version... wasn't there some problems with events on node 6?  This feature seems to be the same with queryParams thing (in hindsight naming of that feature could have been better).

**Comment by elhigu** [comment]
closing as this is already supported.

**Comment by samsch** [comment]
Existing tests for withUserParams don't appear to test anything regarding the data passed to event listeners.

There is also no documentation suggesting that withUserParams would be something usable from event handlers.

I am using current node: 11.4.0.

I can't really create a bug report since I'm not sure what's it's not doing that it *should* do.

Given this code [gist of the files necessary](https://gist.github.com/samsch/8acd479880535091edfbaa68b944b495):
```js
const instance = 13;
const parameterizedKnex = knex.withUserParams({ thisInstance: instance });

parameterizedKnex.on('query-response', (response, obj, builder) => {
  // Where should the userParams be?
  // perhaps builder.client.config?
  console.log('userParams?', builder);
});
parameterizedKnex('things').then(res => console.log('query result', res));
```
Perhaps the userParams should appear at `builder.client.config`, but they don't, and I don't see them anywhere else (a proper JSON.stringify and search for "thisInstance" is thwarted by circular references). If you can suggest what it *should* do, then a more specific reproduction could be made.

**Comment by elhigu** [comment]
> I did attempt to use it, yes, but it didn't appear to be included in the event (`query`, `query-response`) data.

Yes, one improvement for knex's event system that would be nice to have would be to keep reference to the query builder that was used for creating the query so that it could be included in all events . I just looked docs and it says that builder is passed to the query response events, but I need to check that out... I remember that there was actually a problem with that and the documentation might be wrong there. 

Or to pass at least query context of the builder to the events.

> I can't really create a bug report since I'm not sure what's it's not doing that it _should_ do.

I mean using  `.withUserParams` to create a clone and then assigning it a event handler and using closures to access original query builder in the event handler

```
const instance = 13;
const parameterizedKnex = knex.withUserParams({ thisInstance: instance });

parameterizedKnex.on('query-response', (response, obj, builder) => {
  // ofcourse it is not the same thing that passing the final query builder instance that was
  // used to execute the query, but I didn't catch that as an absolute requirement above...
  console.log("user params:", parameterizedKnex.userParams); 
  console.log("other stuff from closure:", instance); 
});
```

I got an impression that something with above didn't work, which would be a bug...

> a proper JSON.stringify and search for "thisInstance" is thwarted by circular references

`console.dir(object, {depth: null, colors: true})` is a nice tool to printout stuff with circular deps.

**Comment by samsch** [comment]
> I got an impression that something with above didn't work, which would be a bug...
That handler will run for any knex query, not just those made with parameterizedKnex.

Perhaps that's a actually a (different) bug? Should event listeners on a cloned instance of knex (via makeKnex or withUserParams) receive events for queries on the original knex or another cloned instance?

**Comment by elhigu** [comment]
There is a test case, which shows that adding event listener to the clone should not change the original instance

https://github.com/tgriesser/knex/blob/master/test/unit/knex.js#L146

but also the listeners of the base instance are copied to the cloned one.

**Comment by samsch** [comment]
The following code still triggers the handler:
```js
const parameterizedKnex1 = knex.withUserParams({ thisInstance: 1 });
const parameterizedKnex2 = knex.withUserParams({ thisInstance: 2 });

parameterizedKnex1.on('query-response', (response, obj, builder) => {
  console.log('parameterizedKnex1 handler');
});

knex('things').then(res => console.log('knex query result', res));
parameterizedKnex2('things').then(res => console.log('parameterizedKnex2 query result', res));
```
Queries run with any clone still trigger listeners on any other. It sounds like perhaps this is not the intended functionality?

**Comment by elhigu** [comment]
which node version are you using?

**Comment by elhigu** [comment]
Ah right you are using 11 :) That does not sound correct I think... maybe those tests are not running correctly. Are you still testing with sqlite like in the gist earlier?

**Comment by samsch** [comment]
Yes, latest `knex` (`0.16.3`) and `sqlite3` (`4.0.6`).
Full file: https://gist.github.com/samsch/1c34f88116a42cb1b101dab89361c3b7

**Comment by kibertoad** [comment]
Foes 0.16.4-next work any better for you?

**Comment by samsch** [comment]
There's no difference using `0.16.4-next1`.

Here's a proper reproduction repo, using 0.16.4-next1. https://github.com/samsch/knex-issue-3047

**Comment by samsch** [comment]
Any confirmation that this isn't (or is) expected behavior?

**Comment by samsch** [comment]
@elhigu If this is the expected behavior, can we re-open this issue, since the functionality doesn't exist? If not, then I can open a new bug issue with the reproduction, and possibly look into fixing it.

**Comment by kibertoad** [comment]
I'll try to find time to look into it on this weekend, sorry for delay!

**Comment by elhigu** [comment]
Re-opening as there might be also a bug involved. Someone should take time to check out if they can reproduce the problem and figure out why current tests are not catching it.

**Comment by samsch** [comment]
So, taking a look through this again, the events do appear to be working properly (handler is only called for queries on the instance it's attached to) on the current master.

I had tested with `0.16.4-next1` available on npm previously, but attempting to dive into this again and figure out the problem, I cloned the master repo and installed from it locally.

Running (a modified version of) the test you linked on the older `0.16.4-next1` does fail, while it passes on master.

Looking forward to 0.16.04 releasing with this fix.
